LibOS/shim/include/shim_fs.h

```diff
@@ -310,31 +334,108 @@ int walk_mounts (int (*walk) (struct shim_mount * mount, void * arg),
 
 /* functions for dcache supports */
 int init_dcache (void);
-int reinit_dcache (void);
 
 extern LOCKTYPE dcache_lock;
 
+/* check permission (specified by mask) of a dentry. If force is not set,
+ * permission is considered granted on invalid dentries */
+/* Assume caller has acquired dcache_lock */
 int permission (struct shim_dentry * dent, int mask, bool force);
 
-int lookup_dentry (struct shim_dentry * base, const char * name, int namelen,
-                   bool force, struct shim_dentry ** new);
-
+/* This function looks up a single dentry based on its parent dentry pointer
+ * and the name.  Namelen is the length of char * name.
+ * The dentry is returned in pointer *new.
+ * 
+ * The force flag causes the libOS to query the underlying file system for the
+ * existence of the dentry, whereas without force, a negative dentry is
+ * treated as definitive.  In some cases, this can be elided as an
+ * optimization.
+ * 
+ * The caller should hold the dcache_lock.
+ */
+int lookup_dentry (struct shim_dentry * parent, const char * name, int namelen,
```

Comment:
This might be clean-uppable.  This argument came in to get all of the different mount/initialization cases working properly.  In particular, I think I had some difficulty with mounting devices that necessitated this.  But there probably is some opportunity for clean up here.